(Mostly) Type-check where clauses on associated types

huonw opened the pull request:
This doesn't yet check that conforming types satisfy the requirements, nor does it do basic sensible-ness checks on the requirements (like `associatedtype T where T == Int`).

huonw committed:
[Parse] Parse trailing where clauses on associated types.

huonw committed:
[AST] Explicitly track the validation state of Decls.

Previously some decls (TypeAliasDecl and ExtensionDecl) had bits
explicitly marking whether they've been validated, while other decls
just deduced this from hasInterfaceType. The doing the latter doesn't
work when the interface type can be computed before doing full
validation (such as protocols and associatedtypes, which have trivial
interface types), and so an explicit bit is adopted for all decls.

huonw committed:
[TypeCheck] Introduce a minimal validation that does enough for name lookup.

There was an underlying recursion between name lookup and validation
which is only exposed by where clauses on associated types, where
looking up an associated type required the protocol to be validated, but
validating a protocol requires any requirements (in where clauses) to be
validated. The latter may of course include references to associated
types, which is problematic. Fortunately, validateDecl is fundamentally
doing more work than necessary for name lookup, so we can break the
recursion by just separating out that minimal work.

At the moment, associated types and protocols are the source of the
recursion, so they're the only ones separated. They just need their
interface types (which is trivial for both) and accessibility (also
simple) for name lookup, along with any inherited protocols for a
protocol.

huonw committed:
[TypeCheck] Explicitly diagnose associatedtype AT = AT.

This warning used to be a consequence of some other warnings, but
changes to how validation work mean this is no longer true.

huonw committed:
[TypeCheck] Diagnose inheriting from invalid things more reliably.

This in-particular is seemingly needed to cover the (rather broken)
following code:

    protocol P { associatedtype AT: AT }

huonw committed:
[TypeCheck] Validate requirements in where clauses on associated types.

huonw committed:
[TypeCheck] Introduce simple ProtocolRequirementTypeResolver.

This does the limited form of resolution required when
validating/resolving types in protocol where clauses.

huonw committed:
[GenericSignatureBuilder] Consume where clauses on associated types.

huonw commented:
Will address the missing pieces and rebase tomorrow.

slavapestov commented:
'associatedtype T where T == Int' could be equivalent to just 'typealias T == Int'

slavapestov commented on lib/AST/SubstitutionMap.cpp:
```diff
@@ -226,18 +226,14 @@ SubstitutionMap
+  auto protocolSelfType = cast<GenericTypeParamType>(
+      protocol->getSelfInterfaceType()->getCanonicalType());
+  SubstitutionMap map;
+  map.addSubstitution(protocolSelfType, selfType);
```
Eventually SubstitutionMaps will be arrays indexed by generic signature requirements; how will this work?

Let's discuss this tomorrow.

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -3240,12 +3240,29 @@ TypeSubstitutionMap TypeBase::getContextSubstitutions(const DeclContext *dc) {
+  // If we're substituting in a protocol, we can avoid the full
+  // generic-signature-based lookup, because the substitutions are easy to
+  // compute. It is important to do this, because name-lookup needs to use this,
+  // and the generic signature of a protocol may not yet be computed.
```
A protocol generic signature is completely trivial though, shouldn't we be able to have one here?

huonw replied:
I'm not sure it is trivial enough: don't all the implied requirements have to be stored too (in the `GenericSignatureBuilder`)?

DougGregor replied:
The signature itself should be trivial `<Self: Proto>`; when we create a `GenericSignatureBuilder` we'll explode out the implied requirements so they can be queried.

To be fair, our nascent plans for storing concrete conformances into the signature will mean that `<Self: Proto>` won't always be trivial. We'll have to think about how to handle that... later.

slavapestov commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -3236,15 +3236,36 @@ ParserResult<TypeDecl> Parser::parseDeclAssociatedType(Parser::ParseDeclOptions
-      return whereStatus;
+    if (Context.isSwiftVersion3()) {
+      // Swift 3 doesn't support where clauses on associated types, so let's
+      // keep exactly the same parsing behaviour as it had.
```
I think we want the new feature to still work in Swift 3 mode too. This was the decision for nested generics and generic subscripts. Only incompatible changes should be limited to Swift 4 mode.

DougGregor replied:
Generally, I think you're right. Although.. use of this feature will completely blow up IRGen, so maybe it's useful as a guard against accidental use/bug reports in the short term.

slavapestov commented on lib/Sema/TypeChecker.h:
```diff
@@ -437,6 +437,11 @@ enum TypeResolutionFlags : unsigned {
+  /// Whether we are checking the outermost layer of types in an inheritance
+  /// clause on something other than an enum (i.e. V, but not U or W, in class
+  /// T: U.V<W>)
+  TR_NonEnumInheritanceClauseOuterLayer = 0x2000000,
```
We need to come up with a better name...

huonw commented:
> 'associatedtype T where T == Int' could be equivalent to just 'typealias T == Int'

Is it worth allowing it just for that purpose, rather than outlawing it and suggesting a `typealias`? (At least, once typealiases work well enough...)

jrose-apple commented:
>> 'associatedtype T where T == Int' could be equivalent to just 'typealias T == Int'
>
> Is it worth allowing it just for that purpose, rather than outlawing it and suggesting a `typealias`? (At least, once typealiases work well enough...)

IIRC this is currently accepted but does not *require* that the typealias be used:

```swift
protocol Base {
  associatedtype Assoc
}
protocol Refined: Base {
  typealias Assoc = Int
}
struct Impl: Refined {
  typealias Assoc = Float
}

func test<T: Base>(_: T) -> T.Assoc? { return nil }
print(type(of: test(Impl()))) // Optional<Float>
```

So at least in Swift 3 mode the associated-type-with-where has a stronger meaning.

DougGregor commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -3236,15 +3236,36 @@ ParserResult<TypeDecl> Parser::parseDeclAssociatedType(Parser::ParseDeclOptions
+        } else {
+          return makeParserCodeCompletionResult<AssociatedTypeDecl>();
+        }
+      }
```
Can we simplify the Swift 3 vs. Swift 4 logic here? Always go down the parsing path, but just drop `TrailingWhere` on the floor with a "this is not implemented yet; use -swift-version 4 to experiment" error of some sort?

DougGregor commented on lib/Sema/GenericTypeResolver.h:
```diff
@@ -144,8 +144,34 @@ class GenericTypeToArchetypeResolver : public GenericTypeResolver {
+///
+/// This should only be used when resolving/validating where clauses in
+/// protocols.
+class ProtocolRequirementTypeResolver : public GenericTypeResolver {
```
It deeply saddens me to have another `GenericTypeResolver`, but I agree that this is the right thing now.

DougGregor commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -1101,7 +1101,11 @@ bool GenericSignatureBuilder::addConformanceRequirement(PotentialArchetype *PAT,
-        continue;
+        if (auto WhereClause = AssocType->getTrailingWhereClause()) {
+          for (auto &req : WhereClause->getRequirements()) {
+            addRequirement(req, &protocolSubMap);
```
I *love* how compact this is!

DougGregor commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -231,6 +267,67 @@ CompleteGenericTypeResolver::recordParamType(ParamDecl *decl, Type type) {
+      return true;
+    }
+
+    // FIXME: Feels too early to perform this check.
```
I know this isn't your FIXME; this should eventually be handled by the generic signature builder, I think.

DougGregor reviewed (approved):
It's looking good!

huonw committed:
[TypeCheck] Check all requirements of a protocol are satisfied by a conformance.

huonw committed:
[AST] Compute the (trivial) generic signature of a protocol on demand.

huonw commented:
@swift-ci Please smoke test.

huonw committed:
[Test] XFAIL two compiler crashers.

These are badly broken and aren't worth my timing to fix now. :(

huonw committed:
[Test] Add two extra errors that shouldn't happen.

I don't have time to completely avoid them. :(

huonw commented:
@swift-ci Please smoke test

huonw commented:
I think I've fixed everything, except for the rather ugly and unfortunate last two commits which just paper over bugs. I don't think I'll get time to address these until I get back, but it still might be useful to land this, for those depending on this feature.

DougGregor commented:
We'll deal with the compiler crashers and get this merged

huonw committed:
[AST] getInterfaceType never returns null, so guard it.

There's no point testing for null on the next line after just calling
getInterfaceType(), because that function asserts that the interface
type exists.

slavapestov commented on lib/AST/GenericSignature.cpp:
```diff
@@ -106,11 +106,11 @@ std::string GenericSignature::gatherGenericParamBindingsText(
 
-    auto found = substitutions.find(canonGP);
-    if (found == substitutions.end())
+    auto type = substitutions(canonGP);
```
It is slightly more correct to do gp.subst(substitutions). Then you don't have to canonicalize it either. But feel free to merge this first and do this later

slavapestov replied:
You can pass in a MakeAbstractConformanceFn.

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -128,6 +128,42 @@ void GenericTypeToArchetypeResolver::recordParamType(ParamDecl *decl, Type type)
 
+Type ProtocolRequirementTypeResolver::resolveGenericTypeParamType(
+    GenericTypeParamType *gp) {
+  assert(gp->getDepth() == 0 && gp->getIndex() == 0 &&
```
Self with non-zero depth does show up in invalid code though... In the past explicit checks like this have always caused problems eventually. You should change this to compare against Proto->getSelfInterfaceType() eventually

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -128,6 +128,42 @@ void GenericTypeToArchetypeResolver::recordParamType(ParamDecl *decl, Type type)
+
+Type ProtocolRequirementTypeResolver::resolveSelfAssociatedType(
+    Type selfTy, AssociatedTypeDecl *assocType) {
+  return assocType->getDeclaredInterfaceType();
```
Assert that 'self' is the protocol's self type then, if you're not substituting here?

slavapestov commented:
@jrose-apple Do you think that's a bug though? If the RHS of the typealias were an associated type, and not concrete, we'd have to 'merge' the two associated types, ie:

    protocol ExpressibleByArrayLiteral { associatedtype Element }
    protocol Sequence { associatedtype Iterator : IteratorProtocol; typealias Element = Iterator.Element }
    protocol ArrayProtocol :  ExpressibleByArrayLiteral, Sequence {}

We really need to infer ExpressibleByArrayLiteral.Element == Sequence.Iterator.Element here.

Also, this example makes no sense at all:

```
protocol P {
  associatedtype A
  typealias B = A
}

protocol Q : P {
  typealias A = Int
}

extension P {
  func foo() {
    print(A.self)
    print(B.self)
  }
}

extension Q {
  func bar() {
    print(A.self)
    print(B.self)
  }
}

struct S1 : Q {
  typealias A = Float
}

print("Conformance of S1 to P: "); S1().foo()
print("Conformance of S1 to Q: "); S1().bar()
```

Note that typealias B is only defined once!

But inside the extension of P, 'B' resolves to the type witness for the associated type A, whereas inside the extension of Q, 'B' resolves to the type alias A, which is always Int.

I feel that the correct behavior is that inside a protocol, 'typealias T = ConcreteType' should always be equivalent to 'associatedtype T where T == ConcreteType', and 'typealias T = AssocType' should always be equivalent to  'associatedtype T where T == AssocType'.

We might be able to hack something up for Swift 3 mode, but it looks more of a consequence of bugs in the generic signature builder than anything intentional.

huonw committed:
[TypeCheck] Bad Self GTPTs from invalid protocols shouldn't crash.

huonw committed:
[TypeCheck] Don't crash with protocol inheritance & AT where clauses.

huonw reviewed (commented):
I realised:

- I forgot to commit/push a fix I had for 3 of the crashers
- I don't know what's going on with the last one, IDE/021, sorry :(
- inherited protocols were crashing (but hopefully not, now)
- somewhat problematically same-type requirements between archetypes seem to disappear from the requirement signature: it looks like `enumerateRequirements` isn't passing them along. This means conformance checking doesn't catch bad things it should (i.e. if there's a `where A.T == B.T` in a protocol declaration, types can conform without this being satisfied), *but* the archetypes seem to still considered equal, so one can still use the types interchangeably in functions (see `needsInherited` in the last commit).

DougGregor commented:
@swift-ci please smoke test

jrose-apple commented:
> I feel that the correct behavior is that inside a protocol, 'typealias T = ConcreteType' should always be equivalent to 'associatedtype T where T == ConcreteType', and 'typealias T = AssocType' should always be equivalent to 'associatedtype T where T == AssocType'.

That's a language change. The current behavior is just like anything defined in a protocol extension: it's shadowable and provides a default for associated types. (Okay, that's a bit of a stretch since typealiases in protocol extensions didn't work right in 3.0 IIRC. But I still don't think it's okay to change the meaning of this in Swift 3 mode.)

DougGregor commented:
`requiresClassSlow()` isn't working properly due to the last commit. I'm testing a fix.

DougGregor commented:
I believe that https://github.com/apple/swift/pull/7799 will address the failures in the standard library.

DougGregor committed:
[Type checker] Validate more minimally before computing accessibility

DougGregor committed:
[Type checker] Dodge crasher when we form a type we cannot print

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor committed:
[Type checker] Minor tweak to crash-avoidance code.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor committed:
Hack for circular associated type issues

DougGregor commented:
Cleanups to come, but I'd like to get this PR merged as-is to keep the history.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test Linux

DougGregor commented:
@swift-ci please smoke test Linux

jrose-apple commented:
I'm sorry I missed this. I'm *very* terrified of `validateDeclForNameLookup`, because I don't think there's any guarantee that we'd come back and finish validating the decl later. What makes this safe?

rjmccall commented:
Nothing. :)  I think that's what I fixed in the second commit of #7852.
